lib/logthrdestdrv.c

```diff
@@ -160,148 +287,150 @@ log_threaded_dest_driver_do_insert(LogThrDestDriver *self)
       result = self->worker.insert(self, msg);
       scratch_buffers_reclaim_marked(mark);
 
-      switch (result)
-        {
-        case WORKER_INSERT_RESULT_DROP:
-          msg_error("Message dropped while sending message to destination",
-                    evt_tag_str("driver", self->super.super.id));
-
-          log_threaded_dest_driver_message_drop(self, msg);
-          _disconnect_and_suspend(self);
-          break;
-
-        case WORKER_INSERT_RESULT_ERROR:
-          self->retries.counter++;
-
-          if (self->retries.counter >= self->retries.max)
-            {
-              if (self->messages.retry_over)
-                self->messages.retry_over(self, msg);
-
-              msg_error("Multiple failures while sending message to destination, message dropped",
-                        evt_tag_str("driver", self->super.super.id),
-                        evt_tag_int("number_of_retries", self->retries.max));
-
-              log_threaded_dest_driver_message_drop(self, msg);
-            }
-          else
-            {
-              log_threaded_dest_driver_message_rewind(self, msg);
-              _disconnect_and_suspend(self);
-            }
-          break;
-
-        case WORKER_INSERT_RESULT_NOT_CONNECTED:
-          log_threaded_dest_driver_message_rewind(self, msg);
-          _disconnect_and_suspend(self);
-          break;
-
-        case WORKER_INSERT_RESULT_REWIND:
-          log_threaded_dest_driver_message_rewind(self, msg);
-          break;
-
-        case WORKER_INSERT_RESULT_SUCCESS:
-          stats_counter_inc(self->written_messages);
-          log_threaded_dest_driver_message_accept(self, msg);
-          break;
-
-        default:
-          break;
-        }
+      self->batch_size++;
+      _process_result(self, result);
 
+      log_msg_unref(msg);
       msg_set_context(NULL);
       log_msg_refcache_stop();
     }
-  if (!self->suspended)
+  if (!self->suspended && self->batch_size > 0 && self->worker.flush)
```

Comment:
This is the right direction, but we will need to be careful in the documentation that flush_lines and flush_bytes are only maximal values, and syslog-ng will flush the first time she can, even if these values are not hit.

This phenomenon might worth reflecting in the configuration: instead of flush_lines and flush_bytes -> batch_mode(on/off), max_lines, max_bytes.